Improve `cleanup` option (code quality, tests, documentation)

ehmicky opened the pull request:
This improves the `cleanup` option, but does not change its behavior:
  - small code simplification
  - added tests for `options.detached: true` (before only `options.detached: false` was tested)
  - fix tests not properly cleaning up created processes (i.e. memory leak)
  - improve option description on `README`

sindresorhus commented on readme.md:
```diff
@@ -277,7 +277,9 @@ Setting this to `false` resolves the promise with the error instead of rejecting
-Keep track of the spawned process and `kill` it when the parent process exits.
+Kill the spawned process when the parent process exits unless either:
+  - the spawned process is [`detached`](https://nodejs.org/api/child_process.html#child_process_options_detached)
+  - the parent process is terminated abruptly, e.g. with `SIGKILL` (as opposed to `SIGTERM` or a normal exit)
```
Don't forget to update the TypeScript definition docs too when doing readme changes.

sindresorhus replied:
```suggestion
  - the parent process is terminated abruptly, for example, with `SIGKILL` as opposed to `SIGTERM` or a normal exit
```

ehmicky commented:
Fixed.

I've also added some tests when the child process exists before the parent process, with the `cleanup` and `detached` options. This was not tested before.

sindresorhus commented on index.d.ts:
```diff
@@ -102,6 +102,7 @@ declare namespace execa {
 
 		/**
 		Keep track of the spawned process and `kill` it when the parent process exits.
+		Kill the spawned process when the parent process exits unless either the spawned process is detached, or the parent process is terminated abruptly.
```
The TS text should be exactly what's in the readme.

ehmicky replied:
Sure, I will copy it.
I was thinking TypeScript comments appear in VS Code "suggestions" as you type, i.e. are read in a different context than a `README` and might need to be more concise. 
But copying the `README` makes it simpler to maintain.

sindresorhus commented on index.d.ts:
```diff
@@ -102,6 +102,7 @@ declare namespace execa {
 
 		/**
 		Keep track of the spawned process and `kill` it when the parent process exits.
```
This was removed from the readme.

ehmicky commented:
Fixed
